core/services/mongodb-remote/src/test/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizerUnitTests.kt

```diff
@@ -1,190 +1,882 @@
-import com.mongodb.stitch.core.services.mongodb.remote.sync.ErrorListener
+import com.mongodb.stitch.core.StitchServiceErrorCode
+import com.mongodb.stitch.core.StitchServiceException
+import com.mongodb.stitch.core.services.mongodb.remote.RemoteDeleteResult
+import com.mongodb.stitch.core.services.mongodb.remote.RemoteUpdateResult
+import com.mongodb.stitch.core.services.mongodb.remote.sync.internal.SyncTestHarness.Companion.withoutVersionId
 import com.mongodb.stitch.server.services.mongodb.local.internal.ServerEmbeddedMongoClientFactory
 import org.bson.BsonDocument
-import org.bson.BsonObjectId
-
-import org.bson.codecs.BsonDocumentCodec
+import org.bson.BsonInt32
 import org.junit.After
+
+import org.junit.Assert.assertEquals
 import org.junit.Assert.assertFalse
+import org.junit.Assert.assertNull
 import org.junit.Assert.assertTrue
-import org.junit.Before
 import org.junit.Test
+import org.mockito.ArgumentCaptor
 import org.mockito.ArgumentMatchers.any
-import org.mockito.Mockito
+import org.mockito.ArgumentMatchers.eq
 import org.mockito.Mockito.`when`
-import org.mockito.Mockito.mock
-import org.mockito.Mockito.spy
 import org.mockito.Mockito.times
 import org.mockito.Mockito.verify
-import java.util.Random
+import java.lang.Exception
 
 class DataSynchronizerUnitTests {
+    companion object {
+        private fun setupPendingReplace(
+            ctx: DataSynchronizerTestContext,
+            expectedDocument: BsonDocument,
+            shouldConflictBeResolvedByRemote: Boolean = false,
+            shouldWaitForError: Boolean = false
+        ) {
+            ctx.mockUpdateResult(RemoteUpdateResult(0, 0, null))
+            ctx.queueConsumableRemoteInsertEvent()
+            ctx.dataSynchronizer.syncDocumentFromRemote(ctx.namespace, ctx.testDocumentId)
+            ctx.doSyncPass()
+
+            // prepare a remote update and a local update.
+            // do a sync pass, accepting the local doc. this will create
+            // a pending replace to be sync'd on the next pass
+            ctx.queueConsumableRemoteUpdateEvent()
+            // set a different update doc than the remote
+            ctx.updateDocument = BsonDocument("\$inc", BsonDocument("count", BsonInt32(2)))
+            ctx.updateTestDocument()
+            // set it back
+            ctx.updateDocument = BsonDocument("\$inc", BsonDocument("count", BsonInt32(1)))
+            ctx.shouldConflictBeResolvedByRemote = shouldConflictBeResolvedByRemote
+
+            ctx.doSyncPass()
+
+            if (shouldWaitForError) {
+                ctx.waitForError()
+            } else {
+                ctx.waitForEvent()
+            }
+
+            val expectedChangeEvent = if (shouldConflictBeResolvedByRemote)
+                ChangeEvent.changeEventForLocalDelete(ctx.namespace, ctx.testDocumentId, false)
+            else ChangeEvent.changeEventForLocalInsert(ctx.namespace, expectedDocument, true)
```

Comment:
[q] Why are we expecting a `changeEventForLocalInsert` instead of a `changeEventForLocalReplace`? This question also applies to everywhere else we expect an insert where we're really expecting a replace event.

Reply:
This event is specific to the circumstances that we are creating here. There are multiple ways to trigger a replace– this happens to be the way I've done it (based on the int tests).